test/e2e/apps/statefulset.go

```diff
@@ -1347,8 +1348,112 @@ var _ = SIGDescribe("StatefulSet", func() {
+
+		e2epv.SkipIfNoDefaultStorageClass(c)
+		ginkgo.By("Creating statefulset " + ssName + " in namespace " + ns)
+		_, err = c.AppsV1().StatefulSets(ns).Create(context.TODO(), ss, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Confirming PVC exists")
+		err = verifyStatefulSetPVCsExist(c, ss, []int{0})
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Confirming Pod is ready")
+		e2estatefulset.WaitForStatusReadyReplicas(c, ss, 1)
+		podName := getStatefulSetPodNameAtIndex(0, ss)
+		pod, err := c.CoreV1().Pods(ns).Get(context.TODO(), podName, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+
+		nodeName := pod.Spec.NodeName
+		framework.ExpectEqual(nodeName, readyNode.Name)
+		node, err := c.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+
+		oldData, err := json.Marshal(node)
+		framework.ExpectNoError(err)
+
+		node.Spec.Unschedulable = true
+
+		newData, err := json.Marshal(node)
+		framework.ExpectNoError(err)
+
+		// cordon node, to make sure pod does not get scheduled to the node until the pvc is deleted
+		patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, v1.Node{})
+		framework.ExpectNoError(err)
+		ginkgo.By("Cordoning Node")
+		_, err = c.CoreV1().Nodes().Patch(context.TODO(), nodeName, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+		framework.ExpectNoError(err)
+		cordoned := true
+
+		defer func() {
+			if cordoned {
+				uncordonNode(c, oldData, newData, nodeName)
+			}
+		}()
+
+		// wait for the node to be unschedulable
+		e2enode.WaitForNodeSchedulable(c, nodeName, 10*time.Second, false)
+
+		ginkgo.By("Deleting Pod")
+		err = c.CoreV1().Pods(ns).Delete(context.TODO(), podName, metav1.DeleteOptions{})
+		framework.ExpectNoError(err)
+
+		// wait for the pod to be recreated
+		e2estatefulset.WaitForStatusCurrentReplicas(c, ss, 1)
+		_, err = c.CoreV1().Pods(ns).Get(context.TODO(), podName, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+
+		pvcList, err := c.CoreV1().PersistentVolumeClaims(ns).List(context.TODO(), metav1.ListOptions{LabelSelector: klabels.Everything().String()})
+		framework.ExpectNoError(err)
+		framework.ExpectEqual(len(pvcList.Items), 1)
+		pvcName := pvcList.Items[0].Name
+
+		ginkgo.By("Deleting PVC")
+		err = c.CoreV1().PersistentVolumeClaims(ns).Delete(context.TODO(), pvcName, metav1.DeleteOptions{})
+		framework.ExpectNoError(err)
+
+		uncordonNode(c, oldData, newData, nodeName)
+		cordoned = false
+
+		ginkgo.By("Confirming PVC recreated")
+		err = verifyStatefulSetPVCsExist(c, ss, []int{0})
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Confirming Pod is ready after being recreated")
+		e2estatefulset.WaitForStatusReadyReplicas(c, ss, 1)
+		pod, err = c.CoreV1().Pods(ns).Get(context.TODO(), podName, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+		framework.ExpectEqual(pod.Spec.NodeName, readyNode.Name) // confirm the pod was scheduled back to the original node
```

Comment:
Does the PVC need to be cleaned up here too?

Reply:
No, since the statefulset will be deleted after the test is done and that will cause the PVC to get deleted

Reply:
Why will deleting the statefulset cause the PVC to be deleted?

(https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/1847-autoremove-statefulset-pvcs is not yet on....)

Reply:
@mattcary sorry, I misunderstood, I have now added the cleanup function to the test, the pvc will now get deleted